### ISSUE TITLE
Mini Cart behaviour "open when adding item" not working on single product page

### ISSUE THREAD

**Issue by vivialice** [woothemes/woocommerce#42638]
## Describe the bug

I have the setting to: "open drawer" when item is added. This works fine on the product grid but when I test it on the single product page, the drawer doesnt open. I've recorded a quick video.  

## To reproduce

1. Set cart to "open drawer"
2. Go to shop page and add item to cart 
3. Go to a product page and add item to the cart

## Expected behavior

Drawer should open any time add to cart button is pressed when this setting is selected. 

## Screenshots

https://user-images.githubusercontent.com/9888201/160683218-0c16340e-894c-4005-bad5-3c72fdc655f9.mov

### WordPress (please complete the following information):

-   Core version: 5.9.2
-   WooCommerce version: 6.3.1
-   WooCommerce Blocks version: 7.3.0
-   Site language: ENG

### Desktop (please complete the following information):

-   OS: MacOS 12.0.1
-   Browser: Firefox
-   Version 98.0.2

**Comment by gigitux** [comment]
For now, we will not work on this issue. Our goal is to have the blockified Single Page template. In this way, we will be able to provide consistent behavior for the Mini Cart block.

I will not close this issue. Consider it as a reminder.

**Comment by onetrev** [comment]
I like the ultimate goal of the blockified Single Page template! In the interim though it feels like the notification "_XYZ item has been added to your cart._" should still appear if you have the mini-cart added to your navigation. Right now it feels broken and the user has very little way to know what happened (other than of course the change to the mini-cart values in the nav).

**Comment by gigitux** [comment]
> I like the ultimate goal of the blockified Single Page template! In the interim though it feels like the notification "_XYZ item has been added to your cart._" should still appear if you have the mini-cart added to your navigation. Right now it feels broken and the user has very little way to know what happened (other than of course the change to the mini-cart values in the nav).

Thanks for sharing your feedback. There is a dedicated issue for this (#6621).

**Comment by katiebethbrown** [comment]
Adding a +1 to note that we ran into this issue using the Mini Cart block on a Team51 site as well, and it wasn't clear that the mini cart not expanding on single product pages when items are added was the expected behavior.

A small note in the block or documentation explaining where this does/doesn't work may be helpful while we wait for the blockified templates.

**Comment by pmcpinto** [comment]
@Aljullu does the estimate (8) refer to opening the drawer on the product page or another solution?

**Comment by Aljullu** [comment]
Yes, this was a tricky one. Currently the _Add to Cart_ button on the Single Product page causes a page refresh, that's why we can't open the Mini Cart drawer on addition. To solve this, we should refactor the logic in WC core so adding to cart from the Single Product page doesn't require a page refresh. After that, making the Mini Cart block open automatically should be fairly easy.

The `8` story points of the estimate should include everything. But I didn't do an in-depth investigation before estimating and considering this needs to be fixed in WC core instead of WC Blocks, I would label this as COU 5, so the estimate is likely not to be very accurate.

**Comment by pmcpinto** [comment]
Thanks for the clarification.
It would be good to do an investigation to reduce the COU and then evaluate if it makes sense to move forward with this.

**Comment by Aljullu** [comment]
Sorry for the delay answering. I just found out that WC core already has an issue for this: https://github.com/woocommerce/woocommerce/issues/36006. IMO, it makes more sense to work on this in the context of WC core instead of WC Blocks because:

1. The changes need to be done in the WC core codebase and, in the end, have nothing to do with the Mini Cart.
2. There are big extensibility concerns that need to be taken into account. While for simple products it might be easy, for other types of products it will probably be a larger effort.

So as the way forward:

1. I'm going to reduce the story points of this issue from 8 → 1, considering it doesn't include the WC core changes.
2. I will mark this issue as blocked by https://github.com/woocommerce/woocommerce/issues/36006.

Please let me know if that doesn't look good! :slightly_smiling_face:

**Comment by pmcpinto** [comment]
Yes, thanks!

Who will solve the Core issue? Origami/Kirigami or the Platform team?

**Comment by Aljullu** [comment]
> Who will solve the Core issue? Origami/Kirigami or the Platform team?

I don't know! :slightly_smiling_face: I think both teams would be good candidates to work on this, so it just depends on prioritization.

**Comment by tomxygen** [comment]
I see this issue and I think the problem is that currently, the Add to cart form block doesn't support Ajax.
That's why, when clicking on add to cart, the page is refreshed.
If you install a plugin like Ajax add to cart for WooCommerce, when clicking on Add to cart, the page won't reload, the product will be added to the cart and the mini cart will slide in from the left just fine.

**Comment by nerrad** [comment]
@pmcpinto, I think this should be tackled by core product teams (not platform) and assessed for priority. IMO it's important to bring consistent behaviour across all instances of the Add to Cart button (especially in a block environment).

**Comment by pmcpinto** [comment]
Good point. Which Core team is the best suited to handle this implementation? cc @chrisdesrochers

**Comment by nerrad** [comment]
Either of Rubik, Kirigami, or Origami. There's overlap for all the teams. It'd probably be good for Origami or Kirigami given their ownership over the Mini Cart block and the connection with that behaviour.

**Comment by pierorocca** [comment]
The mini-cart drawer is an extremely common and popular pattern across top retail sites. It's on trend and more modern than the default notification banner approach.

I'd like to make this the default experience for new Blocks based stores to align with current industry norms. It's makes for a more elevated store experience and it reduces the amount of steps a shopper has to take to get to checkout once they're ready to pay, thus lowering the chance of abandonment.

@ralucaStan as https://github.com/woocommerce/woocommerce/issues/51080 gets completed, this would be a good high impact fix to prioritize given the impact it would have on the overall experience. I left comments here https://github.com/woocommerce/woocommerce/pull/53549#issuecomment-2549465867 on the rationale and provided examples of top websites displaying the mini-cart on a product add. I actually couldn't find a site that didn't use that approach.

**Comment by Aljullu** [comment]
I think this is something that we will be able to fix as we migrate the Add to Cart with Options block to use the iAPI. We are still in early stages, creating the smaller more granular blocks (https://github.com/woocommerce/woocommerce/issues/53014), but in our roadmap we are planning to update those blocks to use the iAPI. This will allow us to Add to Cart without reloading the page, which ultimately will allow the "open when adding item" behavior in the Mini-Cart block.

Based on that, I wonder if this issue should be assigned to Kirigami & Origami instead of Rubik.

**Comment by ResponsiveFlow** [comment]
@Aljullu until the iAPI is implemented, why not allow the mini cart to auto slide in after page reload when a product is added to the cart?
Something similar is done on this store made with WooCommerce but not using blocks:
https://extralandia.it/prodotto/crossword-liner-ultra-matte/

**Comment by pierorocca** [comment]
> @Aljullu until the iAPI is implemented, why not allow the mini cart to auto slide in after page reload when a product is added to the cart?
Something similar is done on this store made with WooCommerce but not using blocks

Good idea. It would need to be much smoother than the example provided. That's quite a jarring experience so it doesn't meet the quality level I had in mind.

**Comment by pierorocca** [comment]
Adding a note from @elizaan36 regarding the fragmentation of the open in drawer FSE setting and the backend redirect to cart setting:

> This is an example where the settings should appear in the admin rather than the Editor since it relates to functionality. Anything that's not appearance related should be removed from editor settings because of this type of confusion.

**Comment by Aljullu** [comment]
> @Aljullu until the iAPI is implemented, why not allow the mini cart to auto slide in after page reload when a product is added to the cart?

Yeah, as an interim approach, it sounds good to me. 👍

**Comment by nerrad** [comment]
> Adding a note from [@elizaan36](https://github.com/elizaan36) regarding the fragmentation of the open in drawer FSE setting and the backend redirect to cart setting:
> 
> > This is an example where the settings should appear in the admin rather than the Editor since it relates to functionality. Anything that's not appearance related should be removed from editor settings because of this type of confusion.

I disagree with this particular case because I think it _is appearance related_ and contextual to what the user is updating (the Mini-cart). It's fine if there's an additional place to modify it (and configuration changes are always synced), however,  beyond being "appearance-related", it's also important to consider what might need to be configured in the context of where the user is working.

**Comment by pierorocca** [comment]
Thanks for the view @nerrad. We have some work to do to align on a guiding principle for 2025, having clear, shared definitions of the terms, and fragmentation traps best avoided. Kudos for pointing out different ways to look at the problem.

I see it as a setting related to the flow, rather than appearance so I agree with Elizabeth's statement that it's functionality. Account creation settings are a good analogous example of a setting that influences where and when the login/signup is presented. 

Let's open this up as a discussion outside of GH. Any consistency and clarity we can achieve will translate into more consistency and clarity for users.

**Comment by pierorocca** [comment]
@ralucaStan I know this is blocked but this is a really important bug to resolve to unblock what is a "must have" UX. See https://github.com/woocommerce/woocommerce/pull/53549#issuecomment-2549465867 for more context. If there are other options besides iAPI in the short term that provides a smooth animation, it's worth a SPIKE.

**Comment by ralucaStan** [comment]
@pierorocca it sounds like #53014 which is WIP will solve this issue. 
 Once #53014 is merged, will there be any instances of single product pages where a page refresh will happen after clicking Add to Card? @Aljullu maybe you can help with an answer

Depending on the above, we can see if it's worth a spike. To be clear, is the spike about opening the drawer after the refresh?

**Comment by ResponsiveFlow** [comment]
@ralucaStan and @pierorocca 
Breakdance implemented exactly this solution.
On page reload the mini cart drawer slides in.

Test it here: https://breakdancelibrary.com/beauty-salon/product/marseille-soap/

**Comment by pierorocca** [comment]
@tomxygen nice! That's a step improvement over an earlier quick POC I saw. The product image refreshes on add to cart or is the entire page refreshing? Is that expected or a limitation? Would it be possible to see it with a theme that has a mini-cart drawer like that on Tsubaki? The larger size makes it easier to observe the UX and the smoothness of the slide relative to on the product catalog page. Thanks for the info. 


![Image](https://github.com/user-attachments/assets/81df1b61-1126-4536-af31-0b687072275c)

@ralucaStan  thanks for sharing #53014. Let's connect separately on the history of that Epic, remaining work, anything else I could benefit being educated on. Appreciate it!

**Comment by Aljullu** [comment]
> Once #53014 is merged, will there be any instances of single product pages where a page refresh will happen after clicking Add to Card?

Yes.

As part of #53014 we are introducing a new _Add to Cart with Options_ block which will use the iAPI. This means it won't require a page refresh when adding products to the cart. However, the current non-iAPI _Add to Cart with Options_ and the _Classic Template_ block (wrappers for PHP templates that require a page refresh) will remain available.

So even though the new block won't require a page refresh when adding to cart, the existing blocks will still require it. We don't plan to force-update existing stores to the new blocks. Hope that makes sense. 🙂

**Comment by pierorocca** [comment]
Awesome thank you @Aljullu. I'm looking forward to this. What's the current ETA for #53014?

**Comment by pierorocca** [comment]
> Adding a note from [@elizaan36](https://github.com/elizaan36) regarding the fragmentation of the open in drawer FSE setting and the backend redirect to cart setting:
> 
> This is an example where the settings should appear in the admin rather than the Editor since it relates to functionality. Anything that's not appearance related should be removed from editor settings because of this type of confusion.
> 
> I disagree with this particular case because I think it is appearance related and contextual to what the user is updating (the Mini-cart). It's fine if there's an additional place to modify it (and configuration changes are always synced), however, beyond being "appearance-related", it's also important to consider what might need to be configured in the context of where the user is working.

Touching on this again @nerrad after a late night reflection thinking about this from either a conceptual model or from a domain model perspective. In the former, the main Concept is products on the details page and the product catalog grid page. Operations on the product Concept include:

- Add to Cart: The primary action to add the product to the shopping cart.
- Show Cart Summary: Displays the cart summary (e.g., mini-cart drawer) when the "Add to Cart" button is clicked.
- Show Cart: Displays the cart page when the "Add to Cart" button is clicked.
- Hide Cart Summary: Provides functionality to hide the cart summary and continue shopping.
- Update Cart Summary: Updates the cart summary information as items are added or removed.
- Update Cart: Modifies the quantity of items already in the cart.

The mini-cart, a noun, is another Concept with its own set of attributes and operations. For example, clicking on the mini-cart "shows cart summary". Once in the drawer, the user could update quantities, remove products, or close the drawer. Or clicking it might instead show a popup summary.

> I disagree with this particular case because I think it is appearance related and contextual to what the user is updating (the Mini-cart)

Coming back to this point and connecting it to conceptual model, the operations "Add to cart", "Show Cart Summary", and "Show Cart" act on the product Concept. Thus IMO settings related to Products, and the Add to Cart operation are directly related and are best grouped together. Operations on the mini-cart and cart as Concept, those to me make sense grouped together be it in the blocks settings or admin setting depending on the choice to separate the various operations and attributes.

This is a long winded way to say in this specific case, "Add to cart" and either showing a notification, redirecting to the cart, or showing the mini-cart, these are user operations on the Product. Product settings are back-end. The benefit of using this type of model is it lends itself to solid logical domain and class design.

Thoughts?

**Comment by nerrad** [comment]
When you approach this from a merchant perspective when they are configuring their store to control shopper functionality: 

- Where would they expect to be able to modify this setting? 
- What indicators would they have about what toggling the option affects?
- Is it a setting that only should be exposed in one place?

I appreciate about how you're trying to group things but I don't think some of the things you describe are mutually exclusive (eg. ""Add to cart" and either showing a notification, redirecting to the cart, or showing the mini-cart, these are user operations on the Product"). I think there's room for thinking what is the default behaviour (controlled at the global level) vs the instance level (controlled at the block level).

Despite this, we need to address the primary issue here first (inconsistent application of the _current_ configuration) and future iterations can be discussed and implemented separately.

**Comment by Aljullu** [comment]
> Awesome thank you @Aljullu. I'm looking forward to this. What's the current ETA for #53014?

@pierorocca We are still in early stages, so this is susceptible to change, but we hope to start testing with extensions in late March/April. At that point, it's a bit unexpected what we will face. While we are approaching the project with backwards compatibility in mind, we fear some extensions might break with the migration to the iAPI or simply by the fact that we will change the markup, class names, etc.

I'm mentioning that because even when we release the block in an experimental mode, we will need to keep in mind it might not be compatible with all extensions out there, so we might not be able to make it the default for some time.

**Comment by pierorocca** [comment]
@Aljullu thanks for the extra insight and the release challenges.

**Comment by pierorocca** [comment]
> When you approach this from a merchant perspective when they are configuring their store to control shopper functionality:

> - Where would they expect to be able to modify this setting?
> - What indicators would they have about what toggling the option affects?
> - Is it a setting that only should be exposed in one place?

@nerrad totally agree we need well thought out principles. If we're not all of the same page, the result is what's in production - a fragmented, confusing experience. 

There could be cases of setting available both globally and locally that work really well. Having a setting for Add to Cart in the mini-cart block while a higher precedence Redirect to Cart setting is only in the back end, is not one of those mutually inclusive cases. It's also challenging to deal with flow altering settings in the block editor. IMO it's harder to configure and visualize flows in the block editor. It's much better as a visual editor and for setting local values. Hopefully we can agree this is a good  example to learn from to instruct future approaches.

**Comment by j111q** [comment]
Jumping into this interesting convo here, thanks @elizaan36 for sharing it with me! 💟 
I'm currently working on the Add to Cart form that @Aljullu mentioned, so this is very timely. 

IMO, this type of setting is more related to the **visual or interactive experience** of a store, and less related to product or store data. So it feels more intuitive to me for it to live in the block editor -- specifically, I'd put it under the `Add to Cart button` settings in the Inspector Panel.

If I were to articulate a broader principle here, it might be something like: `Visual or interactive design choices should be configured in the block editor, while settings relating to store data or product content should be configured in Woo admin.`

As with all our principles, it'll probably have exclusions to the rule, but generally should give us a guideline on where things live. What do you all think though? Does that make sense to you?

If it resonates, I'll add it to our P2 post with the other guidelines (pfrEX1-wi-p2).

**Comment by elizaan36** [comment]
Thanks @j111q - This clarifies things quite well. After a closer look I agree that the setting impacts the _behavior of what’s visible in the canvas_ and is a good fit for the block settings. And as you mentioned, we can align that anything that impacts the actual store data is a better fit for admin settings.

There's still the question of alignment with the admin settings I believe, and improvements to how the block settings <> admin settings interact. @yjailin mentioned that he's taking a look at this next. Since these settings impact all classic themes, backwards compatibility is an issue but we really need to work out how to make this easier to understand. "Enable AJAX add to cart button", for example could be rephrased. Also how to guide the block theme user to configure these settings in the editor instead, or just a simple 1-1 sync?

![Image](https://github.com/user-attachments/assets/0070f104-ebc4-47ba-8eee-dfa7b8bc128e)

**Comment by pierorocca** [comment]
@j111q thanks so much for diving into this. I'll add another example here, apologies for polluting this issue, as a litmus test for the principle. There's a new setting for harmonizing express checkout button styling.

![Image](https://github.com/user-attachments/assets/262b7657-e4ed-4987-bc2f-a6c74a56a1d2)

On the back end, there's no way to know if the block styling override is in place so we are forced to generically say "some settings may be overridden" and the merchant can still make changes on the back end without knowing if they'll have an impact. It's a brutal way of organizing the settings and there's no way to coordinate this well across payment gateways.

![Image](https://github.com/user-attachments/assets/66d441fd-39e7-4582-a27f-48e880bf6356)

If you have any questions about this specifically let's take it to DM so I'm not adding more related but adjacent content to this ticket.

**Comment by elizaan36** [comment]
I think this is an issue that's only in WooPayments? We also have this on the backlog to solve, one option could be to disable the appearance settings in WooPayments when the user is on a block theme and provide direction to the unified setting. But yeah, as you mentioned this is unrelated to the Mini cart issue. Is there a new issue for this @pierorocca or could we create one?

**Comment by pierorocca** [comment]
There's only the feedback that I provided directly to Alex and Pedro @elizaan36, no open ticket . Technically, it would impact all gateways installed on a store with express checkout enabled regardless if they show a notification or not. Changes made on the back end, the user would have no idea they're not affecting anything. Imo it's not a high enough value feature for the complexity it introduces.

**Comment by j111q** [comment]
Adding a screenshot of the relevant bit of my proposed design here for @yjailin's reference when he does start working on this bit which @elizaan36 mentioned: `There's still the question of alignment with the admin settings I believe, and improvements to how the block settings <> admin settings interact. `

<img width="1905" alt="Image" src="https://github.com/user-attachments/assets/36eab1ca-a64a-499a-a383-e5cc24f0c63d" />

Figma file: sUDhknvOsQh19SvIqwNjnA-fi-0_1

**Comment by Aljullu** [comment]
Is the idea to make this setting global, or will it be specific per-block? If the latter, I think one challenge we will have is how we will handle stores that use one block and not the other. For example, what if a store uses the _Mini-Cart_ block but doesn't have the _Add to Cart Button_ block anywhere (either because they use the _Classic Template_ block, the non-blockified _Add to Cart with Options_ block or a custom implementation). And the other way around: if a store is using the _Add to Cart Button_ block, but is not using the _Mini-Cart_ block anywhere, how can we explain this setting won't have any effect?

**Comment by pierorocca** [comment]
@Aljullu good callout. There could be other examples in this transitory state we're in. A consistent way of indicating a setting is not available because of a shortcode <> block incompatibility could add a lot of value. Bonus if the affordance was instantly recognizable as meaning you're using a legacy feature, switch to the new feature for maximum capability.

**Comment by Marc-pi** [comment]
is the Open Drawer after adding a product supposed to work with the new iAPI mini cart, when we click on the Add to Cart Button ?
here on Woo 10.4 it is still not working (TT4 theme / WP 6.9)

<img width="249" height="568" alt="Image" src="https://github.com/user-attachments/assets/45257878-9df7-4edb-b967-471b00d9f8d8" />

**Comment by Aljullu** [comment]
Hey, @Marc-pi, the _Open drawer when adding_ in the Mini-Cart should work when used with the Product Collection or the blockified Add to Cart + Options blocks, as well as when used alongside the product templates from classic themes. In other words, it should work when adding to cart is done without a page refresh.

It doesn't work yet with the non-blockified Add to Cart with Options block, and currently that's not in our top list of priorities, given that we've already released a beta of the blockified Add to Cart + Options block that supports adding to cart without a page refresh. 🙂 

If you are experiencing a different behavior than what I described, then yes, it's likely a bug. Could you open an issue separate from this one with the specific steps to reproduce? 🙏

**Comment by Marc-pi** [comment]
@Aljullu thanks for the clarification 
- i do use the non-blockified Add to Cart with Options block, 
- not the a beta of the blockified Add to Cart + Options block (yet - it is a beta)
- that's the explanation for sure
 
Seems this new block cannot auto detect the product type ? we have to force ? 
How about the case we have several product type in our store (in our case single product and variable products)

<img width="657" height="334" alt="Image" src="https://github.com/user-attachments/assets/4782603f-2786-4853-a550-38d2ebd5d020" />

**Comment by Aljullu** [comment]
Great, thanks for clarifying! In the frontend, the block will detect the product type and display the correct form depending on it.

On the editor side, what you highlighted is the product type switcher. It allows you to edit the form for the different product types, but it doesn't affect which form is rendered on the frontend (the form specific to the product type being rendered will always be used). You can think of it as the view switcher between _Empty_ and _Filed_ cart, it changes what you see in the editor to make it easy to edit the two views, but it doesn't affect the frontend:

<img width="398" alt="Image" src="https://github.com/user-attachments/assets/e3a25207-225b-45c4-8cf4-f5c1629974f3" />

Hope that makes sense! 🙂

**Comment by gigitux** [comment]
@Aljullu Do you think that we can close this issue given that this feature is available when the `Add to Cart + Options block` is used?

**Comment by Aljullu** [comment]
Yes, let's close this. 🙂

**Comment by pierorocca** [comment]
Maybe it's my site, Add to Cart + Options block isn't opening the mini-cart drawer when adding an item to the cart. Are there any known issues or dependencies @Aljullu? I switched blocks using the upgrade notification presented in FSE and can confirm that the "Open drawer when adding" setting is enabled.
